File: packages/domain-tasks/src/filtering/grouping.ts
/**
 * Task list grouping (spec 004 TP015), moved from the renderer's
 * `task-grouping.ts` and `groupTasksByDueDate`.
 *
 * A fixed group carries a **label key** (`dueDate.overdue`, `status.todo`, …)
 * rather than display text, so each surface localizes it; a group named by user
 * data (a project, a folder, a note) carries that `name`. `now` is the caller's.
 */
import { startOfDay } from '../parsing/dates.ts'
import type { Priority, StatusType } from '../parsing/types.ts'
import type { FilterProject, FilterTask, SortDirection, SortField, TaskNoteIndex } from './types.ts'

export interface TaskGroup<T> {
  key: string
  /** Stable key for a fixed label; null when `name` is the label. */
  labelKey: string | null
  /** User data naming the group (project, folder path, note title). */
  name: string | null
  tasks: T[]
  color?: string
  variant?: 'overdue' | 'default'
}

const MS_PER_DAY = 1000 * 60 * 60 * 24

const differenceInDays = (date1: Date, date2: Date): number =>
  Math.floor((date1.getTime() - date2.getTime()) / MS_PER_DAY)

// ============================================================================
// DUE DATE
// ============================================================================

const DUE_DATE_GROUP_ORDER = [
  'overdue',
  'today',
  'tomorrow',
  'upcoming',
  'later',
  'noDueDate'
] as const
type DueDateGroupKey = (typeof DUE_DATE_GROUP_ORDER)[number]

const DUE_DATE_STYLE: Record<DueDateGroupKey, { color?: string; variant?: 'overdue' }> = {
  overdue: { color: '#ef4444', variant: 'overdue' },
  today: { color: '#E5993E' },
  tomorrow: { color: '#3B82F6' },
  upcoming: { color: '#50505A' },
  later: {},
  noDueDate: { color: '#50505A' }
}

/** Buckets by due date relative to `now`, keeping input order within a bucket. */
export const bucketTasksByDueDate = <T extends FilterTask>(
  tasks: readonly T[],
  now: Date
): Record<DueDateGroupKey, T[]> => {
  const groups: Record<DueDateGroupKey, T[]> = {
    overdue: [],
    today: [],
    tomorrow: [],
    upcoming: [],
    later: [],
    noDueDate: []
  }
  const today = startOfDay(now)

  tasks.forEach((task) => {
    if (!task.dueDate) {
      groups.noDueDate.push(task)
      return
    }
    const daysUntil = differenceInDays(startOfDay(task.dueDate), today)
    if (daysUntil < 0) groups.overdue.push(task)
    else if (daysUntil === 0) groups.today.push(task)
    else if (daysUntil === 1) groups.tomorrow.push(task)
    else if (daysUntil <= 7) groups.upcoming.push(task)
    else groups.later.push(task)
  })

  return groups
}

export const groupByDueDate = <T extends FilterTask>(tasks: T[], now: Date): TaskGroup<T>[] => {
  const grouped = bucketTasksByDueDate(tasks, now)
  return DUE_DATE_GROUP_ORDER.map((key) => ({
    key,
    labelKey: `dueDate.${key}`,
    name: null,
    tasks: grouped[key],
    color: DUE_DATE_STYLE[key].color,
    variant: DUE_DATE_STYLE[key].variant
  })).filter((g) => g.tasks.length > 0)
}

// ============================================================================
// PRIORITY
// ============================================================================

const PRIORITY_ORDER: Priority[] = ['urgent', 'high', 'medium', 'low', 'none']

/** Desktop's `priorityConfig[priority].color`. */
export const PRIORITY_COLOR_VARS: Record<Priority, string> = {
  urgent: 'var(--task-priority-urgent)',
  high: 'var(--task-priority-high)',
  medium: 'var(--task-priority-medium)',
  low: 'var(--task-priority-low)',
  none: 'var(--task-priority-none)'
}

export const groupByPriority = <T extends FilterTask>(tasks: T[]): TaskGroup<T>[] => {
  const buckets = new Map<Priority, T[]>(PRIORITY_ORDER.map((p) => [p, []]))

  tasks.forEach((task) => {
    const priority = task.priority || 'none'
    buckets.get(priority)?.push(task)
  })

  return PRIORITY_ORDER.map((priority) => ({
    key: priority,
    labelKey: `priority.${priority}`,
    name: null,
    tasks: buckets.get(priority) ?? [],
    color: PRIORITY_COLOR_VARS[priority]
  })).filter((g) => g.tasks.length > 0)
}

// ============================================================================
// PROJECT
// ============================================================================

export const groupByProject = <T extends FilterTask>(
  tasks: T[],
  projects: readonly FilterProject[]
): TaskGroup<T>[] => {
  const projectMap = new Map(projects.map((p) => [p.id, p]))
  const buckets = new Map<string, T[]>()
  const noProject: T[] = []

  tasks.forEach((task) => {
    const bucket = buckets.get(task.projectId)
    if (bucket) bucket.push(task)
    else if (task.projectId && projectMap.has(task.projectId)) buckets.set(task.projectId, [task])
    else noProject.push(task)
  })

  const result: TaskGroup<T>[] = []
  for (const [projectId, projectTasks] of buckets) {
    const project = projectMap.get(projectId)
    if (!project) continue
    result.push({
      key: projectId,
      labelKey: null,
      name: project.name,
      tasks: projectTasks,
      color: project.color
    })
  }

  result.sort((a, b) => (a.name ?? '').localeCompare(b.name ?? ''))

  if (noProject.length > 0) {
    result.push({ key: 'no-project', labelKey: 'project.none', name: null, tasks: noProject })
  }

  return result
}

// ============================================================================
// CREATED DATE
// ============================================================================

const CREATED_DATE_ORDER = ['today', 'yesterday', 'thisWeek', 'earlier'] as const

const CREATED_DATE_COLORS: Record<(typeof CREATED_DATE_ORDER)[number], string> = {
  today: '#E5993E',
  yesterday: '#3B82F6',
  thisWeek: '#50505A',
  earlier: '#6b7280'
}

export const groupByCreatedDate = <T extends FilterTask>(tasks: T[], now: Date): TaskGroup<T>[] => {
  const today = startOfDay(now)
  const buckets: Record<(typeof CREATED_DATE_ORDER)[number], T[]> = {
    today: [],
    yesterday: [],
    thisWeek: [],
    earlier: []
  }

  tasks.forEach((task) => {
    const daysAgo = differenceInDays(today, startOfDay(task.createdAt))
    if (daysAgo <= 0) buckets.today.push(task)
    else if (daysAgo === 1) buckets.yesterday.push(task)
    else if (daysAgo <= 7) buckets.thisWeek.push(task)
    else buckets.earlier.push(task)
  })

  return CREATED_DATE_ORDER.map((key) => ({
    key,
    labelKey: `createdAt.${key}`,
    name: null,
    tasks: buckets[key],
    color: CREATED_DATE_COLORS[key]
  })).filter((g) => g.tasks.length > 0)
}

// ============================================================================
// STATUS
// ============================================================================

const STATUS_TYPE_ORDER: StatusType[] = ['todo', 'in_progress']

export const groupByStatus = <T extends FilterTask>(
  tasks: T[],
  projects: readonly FilterProject[]
): TaskGroup<T>[] => {
  const statusLookup = new Map<string, { type: StatusType; color: string }>()
  projects.forEach((project) => {
    project.statuses.forEach((status) => {
      if (!statusLookup.has(status.id)) {
        statusLookup.set(status.id, { type: status.type, color: status.color })
      }
    })
  })

  const typeColors = new Map<StatusType, string>()
  for (const { type, color } of statusLookup.values()) {
    if (!typeColors.has(type)) typeColors.set(type, color)
  }

  const buckets = new Map<StatusType, T[]>()
  const uncategorized: T[] = []

  tasks.forEach((task) => {
    const info = statusLookup.get(task.statusId)
    if (!info) {
      uncategorized.push(task)
      return
    }
    if (info.type === 'done') return
    const bucket = buckets.get(info.type)
    if (bucket) bucket.push(task)
    else buckets.set(info.type, [task])
  })

  const result: TaskGroup<T>[] = STATUS_TYPE_ORDER.filter((type) => buckets.has(type)).map(
    (type) => ({
      key: type,
      labelKey: `status.${type}`,
      name: null,
      tasks: buckets.get(type) ?? [],
      color: typeColors.get(type)
    })
  )

  if (uncategorized.length > 0) {
    result.push({
      key: 'uncategorized',
      labelKey: 'status.uncategorized',
      name: null,
      tasks: uncategorized
    })
  }

  return result
}

// ============================================================================
// FOLDER / NOTE
// ============================================================================

/**
 * Group keys are namespaced because the collapsed-group set is stored per tab
 * and shared by every grouping mode. Without the prefix a folder literally
 * named `done` would open collapsed, since `done` is a default-collapsed key.
 */
const FOLDER_GROUP_PREFIX = 'folder-'
const NOTE_GROUP_PREFIX = 'note-'
const ROOT_FOLDER_GROUP_KEY = 'folder-vault-root'
const NO_NOTE_GROUP_KEY = 'no-source-note'

/** `Acme/Legal/NDA` reads as `Acme / Legal / NDA` in a group header. */
const folderGroupLabel = (folderPath: string): string => folderPath.split('/').join(' / ')

/**
 * The note a task is filed under: `sourceNoteId`, else the first related note,
 * else none.
 */
export const getTaskNoteId = (task: FilterTask): string | null =>
  task.sourceNoteId ?? task.linkedNoteIds[0] ?? null

interface NoteBucket<T> {
  folderPath: string
  title: string
  tasks: T[]
}

const bucketTasksByNote = <T extends FilterTask>(
  tasks: T[],
  noteIndex: TaskNoteIndex
): { byNote: Map<string, NoteBucket<T>>; unfiled: T[] } => {
  const byNote = new Map<string, NoteBucket<T>>()
  const unfiled: T[] = []

  tasks.forEach((task) => {
    const noteId = getTaskNoteId(task)
    const info = noteId ? noteIndex.get(noteId) : undefined
    if (!info) {
      unfiled.push(task)
      return
    }
    const bucket = byNote.get(info.id)
    if (bucket) bucket.tasks.push(task)
    else byNote.set(info.id, { folderPath: info.folderPath, title: info.title, tasks: [task] })
  })

  return { byNote, unfiled }
}

export const groupByFolder = <T extends FilterTask>(
  tasks: T[],
  noteIndex: TaskNoteIndex
): TaskGroup<T>[] => {
  const { byNote, unfiled } = bucketTasksByNote(tasks, noteIndex)

  const byFolder = new Map<string, T[]>()
  for (const bucket of byNote.values()) {
    const folderTasks = byFolder.get(bucket.folderPath)
    if (folderTasks) folderTasks.push(...bucket.tasks)
    else byFolder.set(bucket.folderPath, [...bucket.tasks])
  }

  const result: TaskGroup<T>[] = [...byFolder.entries()]
    .filter(([folderPath]) => folderPath !== '')
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([folderPath, folderTasks]) => ({
      key: `${FOLDER_GROUP_PREFIX}${folderPath}`,
      labelKey: null,
      name: folderGroupLabel(folderPath),
      tasks: folderTasks
    }))

  const rootTasks = byFolder.get('')
  if (rootTasks && rootTasks.length > 0) {
    result.push({
      key: ROOT_FOLDER_GROUP_KEY,
      labelKey: 'folder.vaultRoot',
      name: null,
      tasks: rootTasks
    })
  }

  if (unfiled.length > 0) {
    result.push({ key: NO_NOTE_GROUP_KEY, labelKey: 'note.none', name: null, tasks: unfiled })
  }

  return result
}

export const groupByNote = <T extends FilterTask>(
  tasks: T[],
  noteIndex: TaskNoteIndex
): TaskGroup<T>[] => {
  const { byNote, unfiled } = bucketTasksByNote(tasks, noteIndex)

  const result: TaskGroup<T>[] = [...byNote.entries()]
    .sort(
      ([, a], [, b]) => a.folderPath.localeCompare(b.folderPath) || a.title.localeCompare(b.title)
    )
    .map(([noteId, bucket]) => ({
      key: `${NOTE_GROUP_PREFIX}${noteId}`,
      labelKey: null,
      name: bucket.title,
      tasks: bucket.tasks
    }))

  if (unfiled.length > 0) {
    result.push({ key: NO_NOTE_GROUP_KEY, labelKey: 'note.none', name: null, tasks: unfiled })
  }

  return result
}

// ============================================================================
// DISPATCHER
// ============================================================================

export const groupTasksForSort = <T extends FilterTask>(
  tasks: T[],
  sortField: SortField,
  sortDirection: SortDirection,
  projects: readonly FilterProject[],
  now: Date,
  /**
   * Required by the `folder` and `note` modes only. While the notes list is
   * still loading it is undefined, and those modes return no groups so the
   * caller renders the flat list.
   */
  noteIndex?: TaskNoteIndex
): TaskGroup<T>[] => {
  if (sortField === 'title' || sortField === 'completedAt') return []

  let groups: TaskGroup<T>[]

  switch (sortField) {
    case 'dueDate':
      groups = groupByDueDate(tasks, now)
      break
    case 'priority':
      groups = groupByPriority(tasks)
      break
    case 'status':
      groups = groupByStatus(tasks, projects)
      break
    case 'project':
      groups = groupByProject(tasks, projects)
      break
    case 'createdAt':
      groups = groupByCreatedDate(tasks, now)
      break
    case 'folder':
      if (!noteIndex) return []
      groups = groupByFolder(tasks, noteIndex)
      break
    case 'note':
      if (!noteIndex) return []
      groups = groupByNote(tasks, noteIndex)
      break
    default:
      return []
  }

  if (sortDirection === 'desc') {
    groups = [...groups].reverse()
  }

  return groups
}

File: crates/memry-core/src/domain/task_filter/grouping.rs
//! Task list grouping (`packages/domain-tasks/src/filtering/grouping.ts`).
//!
//! A fixed group carries a **label key** (`dueDate.overdue`, `status.todo`, …)
//! rather than display text, so each surface localizes it; a group named by
//! user data (a project, a folder, a note) carries that `name`. Tasks keep
//! their input order inside a group; the caller sorts first.

use std::collections::HashMap;

use super::collation::locale_compare;
use super::config::{SortDirection, SortField};
use super::{FilterProject, FilterTask, Priority, StatusType, TaskNoteInfo};
use crate::domain::calendar::LocalDateTime;

/// One group of the task list, as desktop's `TaskGroup` with task ids.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TaskGroup {
    pub key: String,
    /// Stable key for a fixed label; `None` when `name` is the label.
    pub label_key: Option<String>,
    /// User data naming the group (project name, folder path, note title).
    pub name: Option<String>,
    pub color: Option<String>,
    /// `overdue` for the overdue due-date group, else `None`.
    pub variant: Option<String>,
    pub task_ids: Vec<String>,
}

impl TaskGroup {
    fn fixed(key: &str, label_key: String, tasks: &[&FilterTask], color: Option<&str>) -> Self {
        Self {
            key: key.to_owned(),
            label_key: Some(label_key),
            name: None,
            color: color.map(str::to_owned),
            variant: None,
            task_ids: ids(tasks),
        }
    }

    fn named(key: String, name: String, tasks: &[&FilterTask], color: Option<&str>) -> Self {
        Self {
            key,
            label_key: None,
            name: Some(name),
            color: color.map(str::to_owned),
            variant: None,
            task_ids: ids(tasks),
        }
    }
}

/// `groupTasksForSort`: the groups for a sort field, reversed for `desc`.
///
/// `title`, `completedAt` and an unknown field have no groups. `folder` and
/// `note` need the note index (keyed by note id); without it (the notes list is
/// still loading) they have none, and the caller renders the flat list.
pub fn group_tasks_for_sort(
    tasks: &[&FilterTask],
    field: &SortField,
    direction: &SortDirection,
    projects: &[FilterProject],
    now: LocalDateTime,
    note_index: Option<&HashMap<String, TaskNoteInfo>>,
) -> Vec<TaskGroup> {
    let mut groups = match (field, note_index) {
        (SortField::DueDate, _) => group_by_due_date(tasks, now),
        (SortField::Priority, _) => group_by_priority(tasks),
        (SortField::Status, _) => group_by_status(tasks, projects),
        (SortField::Project, _) => group_by_project(tasks, projects),
        (SortField::CreatedAt, _) => group_by_created_date(tasks, now),
        (SortField::Folder, Some(index)) => group_by_folder(tasks, index),
        (SortField::Note, Some(index)) => group_by_note(tasks, index),
        _ => return Vec::new(),
    };
    if *direction == SortDirection::Desc {
        groups.reverse();
    }
    groups
}

/// `getTaskNoteId`: `sourceNoteId`, else the first linked note, else none.
pub fn task_note_id(task: &FilterTask) -> Option<&str> {
    task.source_note_id
        .as_deref()
        .or_else(|| task.linked_note_ids.first().map(String::as_str))
}

/// Due-date buckets relative to today: `(key, color, variant)`.
const DUE_DATE_GROUPS: [(&str, Option<&str>, Option<&str>); 6] = [
    ("overdue", Some("#ef4444"), Some("overdue")),
    ("today", Some("#E5993E"), None),
    ("tomorrow", Some("#3B82F6"), None),
    ("upcoming", Some("#50505A"), None),
    ("later", None, None),
    ("noDueDate", Some("#50505A"), None),
];

fn group_by_due_date(tasks: &[&FilterTask], now: LocalDateTime) -> Vec<TaskGroup> {
    let today = now.date().days_since_epoch();
    let mut buckets: [Vec<&FilterTask>; 6] = Default::default();
    for &task in tasks {
        let bucket = match task.due_date {
            None => 5,
            Some(due) => match due.date().days_since_epoch() - today {
                ..0 => 0,
                0 => 1,
                1 => 2,
                2..=7 => 3,
                _ => 4,
            },
        };
        buckets[bucket].push(task);
    }
    DUE_DATE_GROUPS
        .iter()
        .zip(buckets)
        .filter(|(_, bucket)| !bucket.is_empty())
        .map(|((key, color, variant), bucket)| TaskGroup {
            variant: variant.map(str::to_owned),
            ..TaskGroup::fixed(key, format!("dueDate.{key}"), &bucket, *color)
        })
        .collect()
}

/// Group order and desktop's `priorityConfig[priority].color`.
const PRIORITY_GROUPS: [(Priority, &str); 5] = [
    (Priority::Urgent, "var(--task-priority-urgent)"),
    (Priority::High, "var(--task-priority-high)"),
    (Priority::Medium, "var(--task-priority-medium)"),
    (Priority::Low, "var(--task-priority-low)"),
    (Priority::None, "var(--task-priority-none)"),
];

fn group_by_priority(tasks: &[&FilterTask]) -> Vec<TaskGroup> {
    PRIORITY_GROUPS
        .iter()
        .filter_map(|&(priority, color)| {
            let bucket: Vec<&FilterTask> = tasks
                .iter()
                .copied()
                .filter(|task| task.priority == priority)
                .collect();
            let name = priority.name();
            (!bucket.is_empty())
                .then(|| TaskGroup::fixed(name, format!("priority.{name}"), &bucket, Some(color)))
        })
        .collect()
}

/// Projects in first-seen order sorted by name; tasks of an empty or unknown
/// project last, under `no-project`.
fn group_by_project(tasks: &[&FilterTask], projects: &[FilterProject]) -> Vec<TaskGroup> {
    // `new Map(projects.map(...))`: a repeated id resolves to the last one.
    let by_id: HashMap<&str, &FilterProject> = projects
        .iter()
        .map(|project| (project.id.as_str(), project))
        .collect();
    let mut buckets: Vec<(&FilterProject, Vec<&FilterTask>)> = Vec::new();
    let mut no_project = Vec::new();
    for &task in tasks {
        if let Some((_, bucket)) = buckets
            .iter_mut()
            .find(|(project, _)| project.id == task.project_id)
        {
            bucket.push(task);
        } else if let Some(project) = by_id
            .get(task.project_id.as_str())
            .filter(|_| !task.project_id.is_empty())
        {
            buckets.push((project, vec![task]));
        } else {
            no_project.push(task);
        }
    }
    buckets.sort_by(|(a, _), (b, _)| locale_compare(&a.name, &b.name));
    let mut groups: Vec<TaskGroup> = buckets
        .into_iter()
        .map(|(project, bucket)| {
            TaskGroup::named(
                project.id.clone(),
                project.name.clone(),
                &bucket,
                Some(&project.color),
            )
        })
        .collect();
    if !no_project.is_empty() {
        groups.push(TaskGroup::fixed(
            "no-project",
            "project.none".to_owned(),
            &no_project,
            None,
        ));
    }
    groups
}

const CREATED_DATE_GROUPS: [(&str, &str); 4] = [
    ("today", "#E5993E"),
    ("yesterday", "#3B82F6"),
    ("thisWeek", "#50505A"),
    ("earlier", "#6b7280"),
];

fn group_by_created_date(tasks: &[&FilterTask], now: LocalDateTime) -> Vec<TaskGroup> {
    let today = now.date().days_since_epoch();
    let mut buckets: [Vec<&FilterTask>; 4] = Default::default();
    for &task in tasks {
        let bucket = match today - task.created_at.date().days_since_epoch() {
            ..=0 => 0,
            1 => 1,
            2..=7 => 2,
            _ => 3,
        };
        buckets[bucket].push(task);
    }
    CREATED_DATE_GROUPS
        .iter()
        .zip(buckets)
        .filter(|(_, bucket)| !bucket.is_empty())
        .map(|((key, color), bucket)| {
            TaskGroup::fixed(key, format!("createdAt.{key}"), &bucket, Some(color))
        })
        .collect()
}

/// `todo` then `in_progress` (done tasks are left out), colored by the first
/// status of that type; tasks whose status resolves in no project last.
fn group_by_status(tasks: &[&FilterTask], projects: &[FilterProject]) -> Vec<TaskGroup> {
    // A status id seen in several projects resolves to the first.
    let mut lookup: HashMap<&str, (StatusType, &str)> = HashMap::new();
    let mut type_colors: Vec<(StatusType, &str)> = Vec::new();
    for status in projects.iter().flat_map(|project| &project.statuses) {
        if lookup.contains_key(status.id.as_str()) {
            continue;
        }
        lookup.insert(&status.id, (status.status_type, &status.color));
        if !type_colors
            .iter()
            .any(|(kind, _)| *kind == status.status_type)
        {
            type_colors.push((status.status_type, &status.color));
        }
    }

    let mut todo = Vec::new();
    let mut in_progress = Vec::new();
    let mut uncategorized = Vec::new();
    for &task in tasks {
        match lookup.get(task.status_id.as_str()) {
            None => uncategorized.push(task),
            Some((StatusType::Todo, _)) => todo.push(task),
            Some((StatusType::InProgress, _)) => in_progress.push(task),
            Some((StatusType::Done, _)) => {}
        }
    }

    let color_of = |kind: StatusType| {
        type_colors
            .iter()
            .find(|(listed, _)| *listed == kind)
            .map(|(_, color)| *color)
    };
    let mut groups = Vec::new();
    for (kind, bucket) in [
        (StatusType::Todo, todo),
        (StatusType::InProgress, in_progress),
    ] {
        if !bucket.is_empty() {
            let name = kind.name();
            groups.push(TaskGroup::fixed(
                name,
                format!("status.{name}"),
                &bucket,
                color_of(kind),
            ));
        }
    }
    if !uncategorized.is_empty() {
        groups.push(TaskGroup::fixed(
            "uncategorized",
            "status.uncategorized".to_owned(),
            &uncategorized,
            None,
        ));
    }
    groups
}

/// Group keys are namespaced because the collapsed-group set is stored per tab
/// and shared by every grouping mode: a folder named `done` must not open
/// collapsed because `done` is a default-collapsed key.
const FOLDER_GROUP_PREFIX: &str = "folder-";
const NOTE_GROUP_PREFIX: &str = "note-";
const ROOT_FOLDER_GROUP_KEY: &str = "folder-vault-root";
const NO_NOTE_GROUP_KEY: &str = "no-source-note";

struct NoteBucket<'a> {
    note: &'a TaskNoteInfo,
    tasks: Vec<&'a FilterTask>,
}

/// Tasks by their resolved note, in first-seen order, and the unfiled rest.
fn bucket_tasks_by_note<'a>(
    tasks: &[&'a FilterTask],
    note_index: &'a HashMap<String, TaskNoteInfo>,
) -> (Vec<NoteBucket<'a>>, Vec<&'a FilterTask>) {
    let mut by_note: Vec<NoteBucket<'a>> = Vec::new();
    let mut unfiled = Vec::new();
    for &task in tasks {
        // An empty id is falsy in the TypeScript and resolves to no note.
        let note = task_note_id(task)
            .filter(|id| !id.is_empty())
            .and_then(|id| note_index.get(id));
        let Some(note) = note else {
            unfiled.push(task);
            continue;
        };
        match by_note.iter_mut().find(|bucket| bucket.note.id == note.id) {
            Some(bucket) => bucket.tasks.push(task),
            None => by_note.push(NoteBucket {
                note,
                tasks: vec![task],
            }),
        }
    }
    (by_note, unfiled)
}

fn group_by_folder(
    tasks: &[&FilterTask],
    note_index: &HashMap<String, TaskNoteInfo>,
) -> Vec<TaskGroup> {
    let (by_note, unfiled) = bucket_tasks_by_note(tasks, note_index);
    let mut by_folder: Vec<(&str, Vec<&FilterTask>)> = Vec::new();
    for bucket in by_note {
        let path = bucket.note.folder_path.as_str();
        match by_folder.iter_mut().find(|(listed, _)| *listed == path) {
            Some((_, folder_tasks)) => folder_tasks.extend(bucket.tasks),
            None => by_folder.push((path, bucket.tasks)),
        }
    }
    let root = by_folder
        .iter()
        .position(|(path, _)| path.is_empty())
        .map(|index| by_folder.remove(index).1);
    by_folder.sort_by(|(a, _), (b, _)| locale_compare(a, b));

    let mut groups: Vec<TaskGroup> = by_folder
        .into_iter()
        .map(|(path, folder_tasks)| {
            TaskGroup::named(
                format!("{FOLDER_GROUP_PREFIX}{path}"),
                path.split('/').collect::<Vec<_>>().join(" / "),
                &folder_tasks,
                None,
            )
        })
        .collect();
    if let Some(root_tasks) = root.filter(|root_tasks| !root_tasks.is_empty()) {
        groups.push(TaskGroup::fixed(
            ROOT_FOLDER_GROUP_KEY,
            "folder.vaultRoot".to_owned(),
            &root_tasks,
            None,
        ));
    }
    push_unfiled(&mut groups, &unfiled);
    groups
}

fn group_by_note(
    tasks: &[&FilterTask],
    note_index: &HashMap<String, TaskNoteInfo>,
) -> Vec<TaskGroup> {
    let (mut by_note, unfiled) = bucket_tasks_by_note(tasks, note_index);
    by_note.sort_by(|a, b| {
        locale_compare(&a.note.folder_path, &b.note.folder_path)
            .then_with(|| locale_compare(&a.note.title, &b.note.title))
    });
    let mut groups: Vec<TaskGroup> = by_note
        .into_iter()
        .map(|bucket| {
            TaskGroup::named(
                format!("{NOTE_GROUP_PREFIX}{}", bucket.note.id),
                bucket.note.title.clone(),
                &bucket.tasks,
                None,
            )
        })
        .collect();
    push_unfiled(&mut groups, &unfiled);
    groups
}

fn push_unfiled(groups: &mut Vec<TaskGroup>, unfiled: &[&FilterTask]) {
    if !unfiled.is_empty() {
        groups.push(TaskGroup::fixed(
            NO_NOTE_GROUP_KEY,
            "note.none".to_owned(),
            unfiled,
            None,
        ));
    }
}

fn ids(tasks: &[&FilterTask]) -> Vec<String> {
    tasks.iter().map(|task| task.id.clone()).collect()
}
